src/redisearch_rs/encode_decode/src/varint/c_api.rs

```diff
@@ -0,0 +1,134 @@
+//! The C API goes here. This part will hopefully all be removed once all users of `varint` API are
+//! oxidized.
+
+use std::ptr::NonNull;
+
+use super::{read, read_field_mask, vector_writer::VectorWriter, write, write_field_mask};
+use crate::{
+    FieldMask,
+    buffer::{BufferReader, BufferWriter, BufferWriterWrapper},
+};
+
+#[unsafe(no_mangle)]
+extern "C" fn ReadVarint(mut b: NonNull<BufferReader>) -> u32 {
+    // Safety: The caller is responsible for ensuring that the pointer is valid.
+    let buffer_reader = unsafe { b.as_mut() };
+    let mut cursor = buffer_reader.as_cursor();
+    let val = read(&mut cursor).unwrap();
+    buffer_reader.pos = cursor.position() as usize;
+
+    val
+}
+
+#[unsafe(no_mangle)]
+// `improper_ctypes_definitions` would be triggered because of u128 crossing the FFI boundary but
+// that's no longer an issue:
+// https://blog.rust-lang.org/2024/03/30/i128-layout-update/#compatibility
+#[allow(improper_ctypes_definitions)]
+extern "C" fn ReadVarintFieldMask(mut b: NonNull<BufferReader>) -> FieldMask {
+    // Safety: The caller is responsible for ensuring that the pointer is valid.
+    let buffer_reader = unsafe { b.as_mut() };
+    let mut cursor = buffer_reader.as_cursor();
+    let val = read_field_mask(&mut cursor).unwrap();
+    buffer_reader.pos = cursor.position() as usize;
+
+    val
+}
+
+/// Note: This function now returns the number of bytes written to the buffer, not the change in
+/// capacity. The change of the buffer capacity is an internal detail and should not be of concern
+/// to the caller.
+#[unsafe(no_mangle)]
+extern "C" fn WriteVarint(value: u32, b: NonNull<BufferWriter>) -> usize {
+    let mut wrapper = BufferWriterWrapper::from(b);
+    write(value, &mut wrapper).unwrap()
+}
+
+/// See the note above for [`WriteVarint`].
+#[unsafe(no_mangle)]
+#[allow(improper_ctypes_definitions)]
+extern "C" fn WriteVarintFieldMask(value: FieldMask, b: NonNull<BufferWriter>) -> usize {
+    let mut wrapper = BufferWriterWrapper::from(b);
+    write_field_mask(value, &mut wrapper).unwrap()
+}
+
+#[unsafe(no_mangle)]
+extern "C" fn NewVarintVectorWriter(cap: usize) -> NonNull<VectorWriter> {
+    let vector_writer = Box::leak(Box::new(VectorWriter::new(cap)));
+
+    // Safety: The pointer is valid because we just created it.
+    unsafe { NonNull::new_unchecked(vector_writer) }
+}
+
+#[unsafe(no_mangle)]
+extern "C" fn VVW_Write(mut writer: NonNull<VectorWriter>, value: u32) -> usize {
+    // Safety: The caller is responsible for ensuring that the pointer is valid.
+    unsafe { writer.as_mut() }.write(value).unwrap()
+}
```

Comment:
Let's be defensive wherever the pointer may end up being null.
Let's get a raw pointer, check (if debug assertions are enabled) if it's not null, and the convert it into a `NonNull` instance (using `new_unchecked` if debug assertions are disabled).

The more breadcrumbs we leave for ourselves when debugging, the better.

Reply:
> Let's be defensive wherever the pointer may end up being null.

But `NonNull` will already panic here if called with a null pointer from the C side. If we're inside the function, we know that the pointer is at least non-null.